pages/1_orders_and_revenue.py

```diff
@@ -18,49 +18,163 @@
+        st.plotly_chart(fig, use_container_width=True)
+
+# Customer Table
+
+# Calculate Total Spend
+total_spend = data.groupby('customer_id')['total_amount'].sum().reset_index()
+
+# Calculate Total Orders
+total_orders = data.groupby('customer_id')['header_id'].nunique().reset_index()
+total_orders.rename(columns={'header_id': 'total_orders'}, inplace=True)
+
+# Calculate Total Refunds
+total_refunds = data.groupby('customer_id')['refund_amount'].sum().reset_index()
+
+# Calculate Last Order Date
+last_order_date = data.groupby('customer_id')['created_at'].max().reset_index()
+
+# Calculate Created Date
+created_date = data.groupby('customer_id')['customer_created_at'].min().reset_index()
+
+# Merge all the calculated fields
+customer_table = data[['customer_id', 'customer_name', 'customer_email', 'customer_city', 'customer_country']].drop_duplicates().reset_index(drop=True)
+customer_table = customer_table.merge(total_spend, on='customer_id', how='left')
+customer_table = customer_table.merge(total_orders, on='customer_id', how='left')
+customer_table = customer_table.merge(total_refunds, on='customer_id', how='left')
+customer_table = customer_table.merge(last_order_date, on='customer_id', how='left')
+customer_table = customer_table.merge(created_date, on='customer_id', how='left')
+
+# Rename columns
+customer_table.rename(columns={
+    'customer_id': 'ID',
+    'customer_name': 'Name',
+    'customer_email': 'Email',
+    'customer_city': 'City',
+    'customer_country': 'Country',
+    'total_amount': 'Total Spend',
+    'refund_amount': 'Total Refunds',
+    'created_at': 'Last Order Date',
+    'customer_created_at': 'Created Date',
+}, inplace=True)
+
+# Format dollar columns
+customer_table['Total Spend'] = customer_table['Total Spend'].apply(lambda x: f"${x:,.2f}")
+customer_table['Total Refunds'] = customer_table['Total Refunds'].apply(lambda x: f"${x:,.2f}")
+
+# Format date columns
+customer_table['Last Order Date'] = customer_table['Last Order Date'].dt.strftime('%Y-%m-%d')
+customer_table['Created Date'] = customer_table['Created Date'].dt.strftime('%Y-%m-%d')
+
+# Display the customer table in Streamlit
+st.markdown("**Enhanced Customer Table**")
+st.dataframe(customer_table)
+
+
+# Location Performance Chart
+# Aggregate revenue by customer_country
+location_performance = data.groupby('customer_country')['total_amount'].sum().reset_index()
+
+# Sort by total_amount in descending order
+location_performance = location_performance.sort_values(by='total_amount', ascending=False)
+
+# Create the Plotly Express choropleth map
+fig = px.choropleth(
+    location_performance,
+    locations='customer_country',
+    locationmode='country names',
+    color='total_amount',
+    color_continuous_scale=px.colors.sequential.Blues,
+    title='Revenue by Country',
+    labels={'customer_country': 'Country', 'total_amount': 'Total Revenue'},
+    hover_name='customer_country',
+    hover_data={'total_amount': True}
+)
```

Comment:
This chart is a bit hard to read on the white background. Here are a few suggestions to apply to make things pop a bit more and also to make the hover components a bit less crowded.

```suggestion
# Sort by total_amount in descending order
location_performance = location_performance.sort_values(by='total_amount', ascending=False)

# Format total_amount for better readability
location_performance['total_amount_formatted'] = location_performance['total_amount'].apply(lambda x: f"${x:,.0f}")

# Define a custom color scale that starts with darker shades of green
custom_color_scale = [
    (0.0, "rgb(198, 219, 239)"),
    (0.2, "rgb(158, 202, 225)"),
    (0.4, "rgb(107, 174, 214)"),
    (0.6, "rgb(66, 146, 198)"),
    (0.8, "rgb(33, 113, 181)"),
    (1.0, "rgb(8, 69, 148)")
]

# Create the Plotly Express choropleth map
fig = px.choropleth(
    location_performance,
    locations='customer_country',
    locationmode='country names',
    color='total_amount',
    color_continuous_scale=custom_color_scale,
    title='Revenue by Country',
    labels={'customer_country': 'Country', 'total_amount_formatted': 'Total Revenue'}
)
```

Reply:
updated-- let me know what you think!

Reply:
Update looks good, thanks!